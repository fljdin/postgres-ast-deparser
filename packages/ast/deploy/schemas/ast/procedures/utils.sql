-- Deploy schemas/ast/procedures/utils to pg

-- requires: schemas/ast/schema

BEGIN;


COMMIT;
