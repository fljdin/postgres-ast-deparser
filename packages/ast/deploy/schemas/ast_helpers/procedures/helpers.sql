-- Deploy schemas/ast_helpers/procedures/helpers to pg

-- requires: schemas/ast_helpers/schema
-- requires: schemas/ast/procedures/types 

BEGIN;


CREATE FUNCTION ast_helpers.coalesce (field text, value text default '')
    RETURNS jsonb
    AS $$
DECLARE
    result jsonb = ast.func_call(
      v_funcname := to_jsonb(ARRAY[ast.string('coalesce')]),
      v_args := to_jsonb(ARRAY[ ast.string(''), ast.a_const(ast.string('')) ])
    );
BEGIN
	result = jsonb_set(result, '{FuncCall, args, 0, String, str}', to_jsonb(field));
	result = jsonb_set(result, '{FuncCall, args, 1, A_Const, String, str}', to_jsonb(value));
	RETURN result;
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.coalesce (field jsonb, value text default '')
    RETURNS jsonb
    AS $$
DECLARE
    result jsonb = ast.func_call(
      v_funcname := to_jsonb(ARRAY[ast.string('coalesce')]),
      v_args := to_jsonb(ARRAY[ ast.string(''), ast.a_const(ast.string('')) ])
    );
BEGIN
	result = jsonb_set(result, '{FuncCall, args, 0}', field);
	result = jsonb_set(result, '{FuncCall, args, 1, A_Const, String, str}', to_jsonb(value));
	RETURN result;
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.tsvectorw (input jsonb, weight text='A')
    RETURNS jsonb
    AS $$
DECLARE
    result jsonb = ast.func_call(
      v_funcname := to_jsonb(ARRAY[ast.string('setweight')]),
      v_args := to_jsonb(ARRAY[input, ast.a_const(ast.string(weight))])
    );
BEGIN
	RETURN result;
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.tsvector (input jsonb)
    RETURNS jsonb
    AS $$
DECLARE
    result jsonb = ast.func_call(
      v_funcname := to_jsonb(ARRAY[ast.string('to_tsvector')]),
      v_args := to_jsonb(ARRAY[input])
    );
BEGIN
	RETURN result;
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.simple_param (
  name text,
  type text
)
    RETURNS jsonb
    AS $$
BEGIN
	RETURN ast.function_parameter(
      v_name := name,
      v_argType := ast.type_name( 
        v_names := to_jsonb(ARRAY[ast.string(type)])
      ),
      v_mode := 105
    );
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.simple_param (
  name text,
  type text,
  default_value text
)
    RETURNS jsonb
    AS $$
BEGIN
	RETURN ast.function_parameter(
      v_name := name,
      v_argType := ast.type_name( 
        v_names := to_jsonb(ARRAY[ast.string(type)])
      ),
      v_mode := 105,
      v_defexpr := ast.string(default_value)
    );
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.simple_param (
  name text,
  type text,
  default_value jsonb
)
    RETURNS jsonb
    AS $$
BEGIN
	RETURN ast.function_parameter(
      v_name := name,
      v_argType := ast.type_name( 
        v_names := to_jsonb(ARRAY[ast.string(type)])
      ),
      v_mode := 105,
      v_defexpr := default_value
    );
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.tsvector (lang text, input jsonb)
    RETURNS jsonb
    AS $$
DECLARE
    result jsonb = ast.func_call(
      v_funcname := to_jsonb(ARRAY[ast.string('to_tsvector')]),
      v_args := to_jsonb(ARRAY[ast.a_const(ast.string(lang)), input])
    );
BEGIN
	RETURN result;
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.a_expr_distinct_tg_field (field text)
    RETURNS jsonb
    AS $$
BEGIN
	RETURN ast.a_expr(v_kind := 3, 
        v_lexpr := ast.column_ref(
          to_jsonb(ARRAY[ ast.string('old'),ast.string(field) ])
        ),
        v_name := to_jsonb(ARRAY[ast.string('=')]),
        v_rexpr := ast.column_ref(
          to_jsonb(ARRAY[ ast.string('new'),ast.string(field) ])
        ) 
    );
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.tsvector_index (fields jsonb)
    RETURNS jsonb
    AS $$
DECLARE
  results jsonb[];
  result jsonb;
  r jsonb;
	i int;
BEGIN
  FOR r IN (select jsonb_array_elements(fields))
    LOOP
      -- TODO maybe we get pg_catalog on some machines from get_current_ts_config and don't need to add it?
      IF ( (r->'lang') IS NULL) THEN
        r = jsonb_set(r, '{lang}', to_jsonb(get_current_ts_config()) );
        -- r = jsonb_set(r, '{lang}', to_jsonb('pg_catalog' || '.' || get_current_ts_config()) );
      END IF;

     -- TODO handle simple/english
      IF ( r->'array' = to_jsonb(true)) THEN
        -- handle array
        results = array_append(results, ast_helpers.tsvectorw( ast_helpers.tsvector(r->>'lang',
          -- start the string
          ast_helpers.coalesce(ast.func_call(
            v_funcname := to_jsonb(ARRAY[ast.string('array_to_string')]),
            v_args := to_jsonb(ARRAY[
              -- type cast null to text[] array
              ast.type_cast(
                v_arg := ast.string(r->>'field'),
                v_typeName := ast.type_name( 
                    v_names := to_jsonb(ARRAY[ast.string(r->>'type')]),
                    v_arrayBounds := to_jsonb(ARRAY[ast.integer(-1)])
                )
              ),
              ast.a_const(ast.string(' '))]
            )
          ))
        -- end array to string function call here
      ) , r->>'weight') );
      ELSE
        IF ( (r->'lang') IS NOT NULL) THEN
          results = array_append(results, ast_helpers.tsvectorw( ast_helpers.tsvector(r->>'lang', ast_helpers.coalesce(r->>'field')) , r->>'weight') );
        ELSE
          -- get_current_ts_config() returns 'english', we'd need to add pg_catalog on there...
          results = array_append(results, ast_helpers.tsvectorw( ast_helpers.tsvector(ast_helpers.coalesce(r->>'field')) , r->>'weight') );
        END IF;
      END IF;
    END LOOP;

  -- create the expressions
  FOR i IN SELECT * FROM generate_subscripts(results, 1) g(i)
    LOOP
      IF (i = 1) THEN
        result = results[i];
      ELSE
        result = ast.a_expr(
          v_kind := 0,
          v_lexpr := results[i], 
          v_name := to_jsonb(ARRAY[ast.string('||')]),
          v_rexpr := result );
      END IF;
    END LOOP;

	RETURN result;
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.create_trigger_with_fields (
  trigger_name text,
  schema_name text,
  table_name text,
  trigger_fn_schema text,
  trigger_fn_name text,
  fields text[],
  timing int default 2,
  events int default 4 | 16
)
    RETURNS jsonb
    AS $$
DECLARE
  results jsonb[];
  result jsonb;
  whenClause jsonb;
  r jsonb;
	i int;
  field text;
BEGIN

  FOR i IN SELECT * FROM generate_subscripts(fields, 1) g(i)
    LOOP
      field = fields[i];
      IF (i = 1) THEN
        whenClause = ast_helpers.a_expr_distinct_tg_field(field);
      ELSE
        whenClause = ast.bool_expr( 
          v_boolop := 1, 
          v_args := to_jsonb(ARRAY[ast_helpers.a_expr_distinct_tg_field(field), whenClause])
        );
      END IF;
    END LOOP;

  result = ast.create_trig_stmt(
    v_trigname := trigger_name,
    v_relation := ast.range_var(schema_name, table_name, true, 'p'),
    v_funcname := to_jsonb(ARRAY[ ast.string(trigger_fn_schema),ast.string(trigger_fn_name) ]),
    v_row := true,
    v_timing := timing,
    v_events := events,
    v_whenClause := whenClause
  );


	RETURN result;
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;


CREATE FUNCTION ast_helpers.create_function (
  schema text,
  name text,
  type text,
  parameters jsonb,
  body text,
  vol text,
  lan text,
  sec int
)
    RETURNS jsonb
    AS $$
DECLARE
  ast jsonb;
BEGIN

  select * FROM ast.create_function_stmt(
    v_funcname := to_jsonb(ARRAY[ ast.string(schema),ast.string(name) ]),
    v_parameters := parameters,
    v_returnType := ast.type_name( 
        v_names := to_jsonb(ARRAY[ast.string(type)])
    ),
    v_options := to_jsonb(ARRAY[
      ast.def_elem(
        'as',
        to_jsonb(ARRAY[ast.string(body)])
      ),
      ast.def_elem(
        'volatility',
        ast.string(vol)
      ),
      ast.def_elem(
        'language',
        ast.string(lan)
      ),
      ast.def_elem(
        'security',
        ast.integer(sec)
      )
    ]::jsonb[])
  ) INTO ast;

  RETURN ast;

END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.create_policy (
  name text,
  vschema text,
  vtable text,
  vrole text,
  qual jsonb,
  cmd text,
  with_check jsonb,
  permissive boolean

)
    RETURNS jsonb
    AS $$
DECLARE
  ast jsonb;
BEGIN
  select * FROM ast.create_policy_stmt(
    v_policy_name := name,
    v_table := ast.range_var(v_schemaname := vschema, v_relname := vtable, v_inh := true, v_relpersistence := 'p'),
    v_roles := to_jsonb(ARRAY[
        ast.role_spec(v_roletype:=0, v_rolename:=vrole)
    ]),
    v_qual := qual,
    v_cmd_name := cmd,
    v_with_check := with_check,
    v_permissive := permissive
  ) INTO ast;
  RETURN ast;
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;


COMMIT;
