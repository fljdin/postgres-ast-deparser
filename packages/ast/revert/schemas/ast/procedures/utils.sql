-- Revert schemas/ast/procedures/utils from pg

BEGIN;

DROP FUNCTION ast.utils;

COMMIT;
