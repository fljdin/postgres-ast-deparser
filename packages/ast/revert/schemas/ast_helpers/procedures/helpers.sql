-- Revert schemas/ast_helpers/procedures/helpers from pg

BEGIN;

DROP FUNCTION ast_helpers.helpers;

COMMIT;
