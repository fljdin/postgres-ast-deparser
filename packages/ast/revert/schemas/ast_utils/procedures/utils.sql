-- Revert schemas/ast_utils/procedures/utils from pg

BEGIN;

DROP FUNCTION ast_utils.utils;

COMMIT;
