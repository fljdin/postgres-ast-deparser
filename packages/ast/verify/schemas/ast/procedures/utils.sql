-- Verify schemas/ast/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast.utils');

ROLLBACK;
