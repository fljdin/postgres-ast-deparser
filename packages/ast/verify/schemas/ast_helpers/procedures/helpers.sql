-- Verify schemas/ast_helpers/procedures/helpers  on pg

BEGIN;

SELECT verify_function ('ast_helpers.helpers');

ROLLBACK;
