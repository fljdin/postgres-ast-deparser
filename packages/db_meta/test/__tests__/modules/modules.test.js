import { getConnections } from '../../utils';
import { initDatabase } from '../../utils/helpers';
import modules from '../../utils/modules';
import { snapshot } from '../../utils/snaps';

let db, dbs, conn, mods, admin, api, sup, teardown;
const objs = {
  tables: {},
  modules: {}
};

const getOutputs = async (newmodule) =>
  (
    await mods.public.select('module_output', ['name', 'value'], {
      module_id: newmodule.id
    })
  ).reduce((m, v) => {
    m[v.name] = v.value;
    return m;
  }, {});

const checkOutputs = async (newmodule) => {
  const outputs = await getOutputs(newmodule);
  expect(snapshot(outputs)).toMatchSnapshot();
};

const installModule = async (name, data) => {
  const [newmodule] = await mods.public.callAny('install_module', {
    database_id: objs.database1.id,
    module_defn_id: objs.modules[name].id,
    context: 'sql',
    data
  });
  return newmodule;
};

beforeAll(async () => {
  ({ db, conn, teardown } = await getConnections());
  await db.begin();
  await db.savepoint('db');
  dbs = db.helper('collections_public');
  await initDatabase({ objs, dbs, dbname: 'simpledb' });

  admin = {
    public: db.helper('collections_public'),
    private: db.helper('collections_private')
  };
  mods = {
    public: db.helper('modules_public'),
    private: db.helper('modules_private')
  };

  await modules({ objs, mods });

  api = {
    public: conn.helper(objs.database1.schema_name),
    private: conn.helper(objs.database1.private_schema_name)
  };
  sup = {
    public: db.helper(objs.database1.schema_name),
    private: db.helper(objs.database1.private_schema_name)
  };
});

afterAll(async () => {
  try {
    await db.rollback('db');
    await db.commit();
    await teardown();
  } catch (e) {
    console.log(e);
  }
});

beforeEach(async () => {
  await db.savepoint('each');
});
afterEach(async () => {
  await db.rollback('each');
});

it('list deps', async () => {
  const deps = await mods.private.callAny('get_all_module_deps', {
    module_defn_id: objs.modules.crypto_auth_module.id
  });
  const ds = deps.map((dep) => dep.name).sort();
  expect(ds).toMatchSnapshot();
});

it('cannot install w/o deps', async () => {
  await db.savepoint('error1');
  let failed = false;
  let message = '';
  try {
    await mods.public.insertOne('modules', {
      database_id: objs.database1.id,
      module_defn_id: objs.modules.tokens_module.id,
      context: 'sql'
    });
  } catch (e) {
    message = e.message;
    failed = true;
  }
  expect(failed).toBe(true);
  expect(message).toEqual('MISSING_REQUIRED_MODULE: users_module');
  await db.rollback('error1');
});

it('immutable_field_utils_module', async () => {
  const mod = await installModule('immutable_field_utils_module');
  await checkOutputs(mod);
});

it('immutable_field_module', async () => {
  await installModule('immutable_field_utils_module');
  const tbl = await admin.public.insertOne('table', {
    database_id: objs.database1.id,
    schema_id: objs.database1.schemas.public.id,
    name: 'immutable_fields_table'
  });
  await admin.public.insertOne('field', {
    table_id: tbl.id,
    name: 'id',
    type: 'serial'
  });
  const value_on_insert = await admin.public.insertOne('field', {
    table_id: tbl.id,
    name: 'value_on_insert',
    type: 'text'
  });
  const null_on_insert = await admin.public.insertOne('field', {
    table_id: tbl.id,
    name: 'null_on_insert',
    type: 'text'
  });

  await installModule('immutable_field_module', {
    field_id: value_on_insert.id
  });

  await installModule('immutable_field_module', {
    field_id: null_on_insert.id
  });

  await sup.public.insertOne('immutable_fields_table', {
    value_on_insert: 'test-immutable',
    null_on_insert: null
  });

  // NOTE
  // we may still need the verify function

  const rows1 = await sup.public.select('immutable_fields_table', ['*']);
  expect(rows1).toMatchSnapshot();
  await sup.public.update(
    'immutable_fields_table',
    {
      null_on_insert: 'updated-value'
    },
    {
      id: 1
    }
  );
  const rows2 = await sup.public.select('immutable_fields_table', ['*']);
  expect(rows2).toMatchSnapshot();
  let fail = false;
  let message;
  await db.savepoint('error2');
  try {
    await sup.public.update(
      'immutable_fields_table',
      {
        null_on_insert: 'updated-value2'
      },
      {
        id: 1
      }
    );
  } catch (e) {
    fail = true;
    message = e.message;
  }
  expect(fail).toBe(true);
  expect(message).toEqual('IMMUTABLE_PROPERTY null_on_insert');
  await db.rollback('error2');
  await db.savepoint('error3');
  fail = false;
  message = '';
  try {
    await sup.public.update(
      'immutable_fields_table',
      {
        value_on_insert: 'updated-value2'
      },
      {
        id: 1
      }
    );
  } catch (e) {
    fail = true;
    message = e.message;
  }
  expect(fail).toBe(true);
  expect(message).toEqual('IMMUTABLE_PROPERTY value_on_insert');
  await db.rollback('error3');
});

it('encrypted_secrets_utils_module', async () => {
  const mod = await installModule('encrypted_secrets_utils_module');
  await checkOutputs(mod);
});

it('encrypted_secrets_encrypt_field_module', async () => {
  await installModule('encrypted_secrets_utils_module');
  /*

NOTES:

currently it's BEST if

* crypt uses text fields
* pgp uses bytea fields

** pgp can technically use both, but bytea is more "Natural"

** crypt as bytea will not be able to verify yet until we make a verify fn that uses

    prv.encrypt_field_bytea_to_text(crypt_bytea_field) =
    crypt($1::bytea::text, prv.encrypt_field_bytea_to_text(crypt_bytea_field)) as result

WHICH IS WAYYY TO COMPLEX.

So, for now, just force crypt = text, pgp = bytea ;)

*/

  const tbl = await admin.public.insertOne('table', {
    database_id: objs.database1.id,
    schema_id: objs.database1.schemas.public.id,
    name: 'new_secrets_tbl'
  });
  const pgp_field = await admin.public.insertOne('field', {
    table_id: tbl.id,
    name: 'pgp_field',
    type: 'bytea'
  });
  const pgp_text_field = await admin.public.insertOne('field', {
    table_id: tbl.id,
    name: 'pgp_text_field',
    type: 'text'
  });
  const crypt_field = await admin.public.insertOne('field', {
    table_id: tbl.id,
    name: 'crypt_field',
    type: 'text'
  });
  const crypt_bytea_field = await admin.public.insertOne('field', {
    table_id: tbl.id,
    name: 'crypt_bytea_field',
    // type: 'bytea'
    type: 'text'
  });
  const owner_id = await admin.public.insertOne('field', {
    table_id: tbl.id,
    name: 'owner_id',
    type: 'uuid'
  });

  await installModule('encrypted_secrets_getter_module', {
    table_id: tbl.id,
    owned_field: 'owner_id'
  });

  await installModule('encrypted_secrets_setter_module', {
    table_id: tbl.id,
    owned_field: 'owner_id'
  });

  await installModule('encrypted_secrets_verify_module', {
    table_id: tbl.id,
    owned_field: 'owner_id'
  });

  await installModule('encrypted_secrets_encrypt_field_module', {
    field_id: pgp_field.id,
    encode_field_id: owner_id.id,
    encryption_type: 'pgp'
  });

  await installModule('encrypted_secrets_encrypt_field_module', {
    field_id: pgp_text_field.id,
    encode_field_id: owner_id.id,
    encryption_type: 'pgp'
  });

  await installModule('encrypted_secrets_encrypt_field_module', {
    field_id: crypt_field.id,
    encryption_type: 'crypt'
  });

  await installModule('encrypted_secrets_encrypt_field_module', {
    field_id: crypt_bytea_field.id,
    encryption_type: 'crypt'
  });

  await sup.public.insertOne('new_secrets_tbl', {
    pgp_field: 'hello-encryption',
    pgp_text_field: 'hello-encryption',
    crypt_field: 'hello-encryption',
    crypt_bytea_field: 'hello-encryption',
    owner_id: objs.database1.id // any uuid
  });

  const [{ result: res1 }] = await db.any(
    `SELECT
    crypt_field = crypt($1, crypt_field) as result
       FROM
       "${objs.database1.schema_name}".new_secrets_tbl`,
    ['hello-encryption']
  );

  expect(res1).toBe(true);

  // const [{result: res2}] = await db.any(`SELECT
  // "${objs.database1.private_schema_name}".encrypt_field_bytea_to_text(crypt_bytea_field) = crypt($1::bytea::text, "${objs.database1.private_schema_name}".encrypt_field_bytea_to_text(crypt_bytea_field)) as result
  //    FROM
  //    "${objs.database1.schema_name}".new_secrets_tbl`, ['hello-encryption']);

  // expect(res2).toBe(true);

  const [{ encrypt_field_pgp_get: pgpres1 }] = await db.any(`SELECT
      "${objs.database1.private_schema_name}".encrypt_field_pgp_get(pgp_field, owner_id::text)
       FROM
       "${objs.database1.schema_name}".new_secrets_tbl`);

  expect(pgpres1).toMatchSnapshot();

  const [{ encrypt_field_pgp_get: pgpres2 }] = await db.any(`SELECT
      "${objs.database1.private_schema_name}".encrypt_field_pgp_get(pgp_text_field::bytea, owner_id::text)
       FROM
       "${objs.database1.schema_name}".new_secrets_tbl`);

  expect(pgpres2).toMatchSnapshot();
});

it('default_ids_module', async () => {
  const mod = await installModule('default_ids_module');
  await getOutputs(mod);
});

it('create_table_module', async () => {
  await installModule('default_ids_module');
  const mod = await installModule('create_table_module', {
    table_name: 'table_with_uuid'
  });
  await getOutputs(mod);
});

it('create_serial_table_module', async () => {
  await installModule('default_ids_module');
  const mod = await installModule('create_table_module', {
    table_name: 'table_with_serial',
    pk_type: 'serial'
  });
  await sup.public.insertOne('table_with_serial');
  await sup.public.insertOne('table_with_serial');
  const val = await sup.public.insertOne('table_with_serial');
  expect(val).toMatchSnapshot();
  await getOutputs(mod);
});

it('uuid_module', async () => {
  const newmodule = await installModule('uuid_module');
  await checkOutputs(newmodule);
});

it('users_modules', async () => {
  const newmodule = await installModule('users_module', {
    table_name: 'users'
  });
  await checkOutputs(newmodule);
});

it('secrets_module', async () => {
  const usermodule = await installModule('users_module', {
    table_name: 'users'
  });
  const userouts = await getOutputs(usermodule);
  const secretsmodule = await installModule('secrets_module', {
    secrets_table: 'user_secrets',
    owned_table_id: userouts.table_id
  });
  await checkOutputs(secretsmodule);
});

it('tokens_module', async () => {
  await installModule('users_module');
  const tokensmod = await installModule('tokens_module');
  await checkOutputs(tokensmod);
});

it('crypto_module', async () => {
  const usermodule = await installModule('users_module', {
    table_name: 'users'
  });
  const userouts = await getOutputs(usermodule);
  await installModule('tokens_module');
  await installModule('secrets_module', {
    owned_table_id: userouts.table_id,
    secrets_table: 'user_secrets'
  });
  const cryptomod = await installModule('crypto_auth_module', {
    user_field: 'bitg_address',
    crypto_network: 'BITG',
    sign_up_unique_key: 'sign_up_with_bitg_address'
  });
  await checkOutputs(cryptomod);
});

it('rls', async () => {
  await installModule('users_module');
  await installModule('tokens_module');
  const rlsmod = await installModule('rls_module');
  await checkOutputs(rlsmod);
});

it('peoplestamps', async () => {
  await installModule('users_module');
  await installModule('tokens_module');
  await installModule('rls_module');
  const mod = await installModule('peoplestamps_module');
  await checkOutputs(mod);
});

it('peoplestamps table', async () => {
  await installModule('users_module');
  await installModule('tokens_module');
  await installModule('rls_module');
  await installModule('peoplestamps_module');
  const tbl = await admin.public.insertOne('table', {
    database_id: objs.database1.id,
    schema_id: objs.database1.schemas.public.id,
    name: 'peoplestamped_table'
  });
  const mod = await installModule('table_peoplestamps_module', {
    table_id: tbl.id
  });
  await checkOutputs(mod);
});

it('timestamps', async () => {
  const mod = await installModule('timestamps_module');
  await checkOutputs(mod);
});

it('timestamps table', async () => {
  await installModule('timestamps_module');
  const tbl = await admin.public.insertOne('table', {
    database_id: objs.database1.id,
    schema_id: objs.database1.schemas.public.id,
    name: 'timestamped_table'
  });
  const mod = await installModule('table_timestamps_module', {
    table_id: tbl.id
  });
  await checkOutputs(mod);
});

it('emails_module', async () => {
  await installModule('users_module');
  const mod = await installModule('emails_module');
  await checkOutputs(mod);
});

it('encrypted_secrets_module', async () => {
  await installModule('users_module');
  const mod = await installModule('encrypted_secrets_module');
  await checkOutputs(mod);
});

it('user_auth_module', async () => {
  await installModule('jobs_module');
  const usermod = await installModule('users_module');
  await installModule('tokens_module');
  await installModule('rls_module');
  const userouts = await getOutputs(usermod);
  await installModule('secrets_module', {
    secrets_table: 'user_secrets',
    owned_table_id: userouts.table_id
  });
  await installModule('encrypted_secrets_module');
  await installModule('emails_module');

  const mod = await installModule('user_auth_module');
  await checkOutputs(mod);
});

it('jobs_module', async () => {
  const mod = await installModule('jobs_module');
  await checkOutputs(mod);
});

it('jobs_trigger_module', async () => {
  await installModule('jobs_module');

  const tbl = await admin.public.insertOne('table', {
    database_id: objs.database1.id,
    schema_id: objs.database1.schemas.public.id,
    name: 'jobs_trigger_table'
  });

  const mod = await installModule('jobs_trigger_module', {
    table_id: tbl.id,
    job_name: 'my-custom-fn'
  });
  await checkOutputs(mod);
});

it('invites_module', async () => {
  await installModule('jobs_module');
  await installModule('users_module');
  await installModule('emails_module');
  await installModule('tokens_module');
  await installModule('rls_module');
  const mod = await installModule('invites_module');
  await checkOutputs(mod);
});
