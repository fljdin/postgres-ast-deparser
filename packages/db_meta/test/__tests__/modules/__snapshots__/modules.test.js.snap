// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create_serial_table_module 1`] = `
Object {
  "id": 3,
}
`;

exports[`crypto_module 1`] = `
Object {
  "crypto_network": "BITG",
  "schema": "[schemahash]",
  "sign_in_record_failure": "sign_in_record_failure",
  "sign_in_request_challenge": "sign_in_request_challenge",
  "sign_in_with_challenge": "sign_in_with_challenge",
  "sign_up_unique_key": "sign_up_with_bitg_address",
  "user_field": "bitg_address",
}
`;

exports[`emails_module 1`] = `
Object {
  "emails_table": "emails",
  "table_id": "[ID]",
}
`;

exports[`encrypted_secrets_encrypt_field_module 1`] = `"hello-encryption"`;

exports[`encrypted_secrets_encrypt_field_module 2`] = `"hello-encryption"`;

exports[`encrypted_secrets_module 1`] = `
Object {
  "encrypt_schema": "[schemahash]",
  "secrets_table": "encrypted_secrets",
  "table_id": "[ID]",
  "upsert_schema": "[schemahash]",
}
`;

exports[`encrypted_secrets_utils_module 1`] = `
Object {
  "encrypted_schema": "[schemahash]",
}
`;

exports[`immutable_field_module 1`] = `
Array [
  Object {
    "id": 1,
    "null_on_insert": null,
    "value_on_insert": "test-immutable",
  },
]
`;

exports[`immutable_field_module 2`] = `
Array [
  Object {
    "id": 1,
    "null_on_insert": "updated-value",
    "value_on_insert": "test-immutable",
  },
]
`;

exports[`immutable_field_utils_module 1`] = `
Object {
  "trigger_function": "immutable_field_tg",
  "trigger_schema": "[schemahash]",
}
`;

exports[`invites_module 1`] = `
Object {
  "claimed_invites_table": "claimed_invites",
  "invites_schema": "[schemahash]",
  "invites_table": "invites",
}
`;

exports[`jobs_module 1`] = `
Object {
  "jobs_schema": "[schemahash]",
}
`;

exports[`jobs_trigger_module 1`] = `
Object {
  "event_type": "INSERT",
  "table_id": "[ID]",
  "trigger_function": "tg_add_job_with_row",
  "trigger_name": "jobs_trigger_table_insert_job_tg",
}
`;

exports[`list deps 1`] = `
Array [
  "crypto_auth_module",
  "secrets_module",
  "tokens_module",
  "users_module",
]
`;

exports[`peoplestamps 1`] = `
Object {
  "created_by_column": "updated_by",
  "trigger_name": "tg_peoplestamps",
  "trigger_schema": "[schemahash]",
  "updated_by_column": "created_by",
}
`;

exports[`peoplestamps table 1`] = `Object {}`;

exports[`rls 1`] = `
Object {
  "authenticate": "authenticate",
  "authenticate_schema": "[schemahash]",
  "current_group_ids": "current_group_ids",
  "current_role": "current_user",
  "current_role_id": "[ID]",
  "role_schema": "[schemahash]",
}
`;

exports[`secrets_module 1`] = `
Object {
  "secrets_schema": "[schemahash]",
  "secrets_table": "user_secrets",
  "table_id": "[ID]",
}
`;

exports[`timestamps 1`] = `
Object {
  "created_at_column": "created_at",
  "trigger_name": "tg_timestamps",
  "trigger_schema": "[schemahash]",
  "updated_at_column": "updated_at",
}
`;

exports[`timestamps table 1`] = `Object {}`;

exports[`tokens_module 1`] = `
Object {
  "table_id": "[ID]",
}
`;

exports[`user_auth_module 1`] = `
Object {
  "auth_schema": "[schemahash]",
  "forgot_password": "forgot_password",
  "reset_password": "reset_password",
  "send_verification_email": "send_verification_email",
  "set_password": "set_password",
  "sign_in": "login",
  "sign_up": "register",
  "verify_email": "verify_email",
}
`;

exports[`users_modules 1`] = `
Object {
  "table_id": "[ID]",
}
`;

exports[`uuid_module 1`] = `
Object {
  "seeded_uuid_function": "uuid_generate_v4",
  "seeded_uuid_function_with_args": "uuid_generate_seeded_uuid",
  "seeded_uuid_related_trigger": "seeded_uuid_related_trigger",
  "uuid_schema": "[schemahash]",
}
`;
